### ISSUE TITLE
EU Omnibus Directive: Genuine Reviews

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#4812]
For sellers in/into the EU, this comes into force 28 May 2022.
This article is as good as any:
https://www.loganpartners.com/the-omnibus-directive-consumer-reviews/

Currently, it is possible in ZC to force users to be logged in to write a review, but to provide a guarantee that they are also users of the product in question, they should also be restricted to only those products previously purchased.

Additionally some text should be displayed to publicly state the restrictions that are in place, as part of what the shop owner is doing to ensure reviews are genuine.

A top of the head suggestion of the best approach for this restriction would be of value, prior to doing some hacklet.

**Comment by torvista** [comment]
I did this in an additional header file, but maybe this functionality is a candidate for adding to the Customer class?

```
if (zen_is_logged_in()) {
    $sql = "SELECT op.orders_id FROM " . TABLE_ORDERS_PRODUCTS . " op 
               LEFT JOIN " . TABLE_ORDERS . " o ON o.orders_id = op.orders_id 
               WHERE op.products_id = :productsId:  
               AND o.customers_id = :customerId: 
               ORDER BY o.orders_id LIMIT 1";
    $sql = $db->bindVars($sql, ':productsId:', $_GET['products_id'], 'integer');
    $sql = $db->bindVars($sql, ':customerId:', $_SESSION['customer_id'], 'integer');
    $results = $db->Execute($sql);
    $flag_allow_review_write = $results->count() > 0;
}
```

**Comment by g2ktcf** [comment]
While I understand the requirement, this fix means that users who bought the product at another site cannot comment.  Is that the intent of the rule?  What if the order was canceled? Returned? What about those confused customers who create multiple accounts? 

> For example, these steps could include requesting information to verify that that the consumer has actually used or purchased the product, or requiring reviewers to register on a specific platform. However, traders should avoid taking measures which would make it excessively difficult for consumers to post reviews thus discouraging real consumers from submitting reviews.

For the part below, we could tag the review as "pending" if they did not purchase it on the site and ask for follow on information to prove that they are indeed USERS of that product/service.  

While this does not affect me now, I see this happening all over due to fraud.

**Comment by torvista** [comment]
My "solution" was just so I could put some corresponding text next to the reviews to state how reviews are restricted to genuine purchasers. Extrapolating that from my little local store out to all the possible what-ifs and the global manipulation of reviews is way out of my interest and knowledge.
It is not meant to be a fix, but could save someone some time as a quick fix.

**Comment by g2ktcf** [comment]
@torvista  yes, I was thinking in a comprehensive way.  I tend to think this way..."how could this go wrong,"  The solution you propose is definitely a step forward!  I do not see any negatives to including it for all.
